ui/src/routes/devices.$id.settings.macros.tsx

```diff
@@ -0,0 +1,255 @@
+import { useEffect, Fragment, useMemo, useState } from "react";
+import { useNavigate } from "react-router-dom";
+import { LuPenLine, LuLoader, LuCopy, LuMoveRight, LuCornerDownRight, LuArrowUp, LuArrowDown } from "react-icons/lu";
+
+import { KeySequence, useMacrosStore, generateMacroId } from "@/hooks/stores";
+import { SettingsPageHeader } from "@/components/SettingsPageheader";
+import { Button } from "@/components/Button";
+import EmptyCard from "@/components/EmptyCard";
+import Card from "@/components/Card";
+import { MAX_TOTAL_MACROS, COPY_SUFFIX } from "@/constants/macros";
+import { keyDisplayMap, modifierDisplayMap } from "@/keyboardMappings";
+import notifications from "@/notifications";
+import { SettingsItem } from "@/routes/devices.$id.settings";
+
+const normalizeSortOrders = (macros: KeySequence[]): KeySequence[] => {
+  return macros.map((macro, index) => ({
+    ...macro,
+    sortOrder: index + 1,
+  }));
+};
+
+export default function SettingsMacrosRoute() {
+  const { macros, loading, initialized, loadMacros, saveMacros } = useMacrosStore();
+  const navigate = useNavigate();
+  const [actionLoadingId, setActionLoadingId] = useState<string | null>(null);
+  
+  const isMaxMacrosReached = useMemo(() => 
+    macros.length >= MAX_TOTAL_MACROS, 
+    [macros.length]
+  );
+
+  useEffect(() => {
+    if (!initialized) {
+      loadMacros();
+    }
+  }, [initialized, loadMacros]);
+
+  const handleDuplicateMacro = async (macro: KeySequence) => {
```

Comment:
You're dependent on react state. Wrap in `useCallback`

Reply:
added, i think it might be over optimisation though